### PR TITLE
Refactor Barman scripts and add a sub-command to switch Barman config

### PR DESCRIPTION
**Pull request by barthisrael**
We currently have a script named `patroni_barman_recover` in Patroni, which is intended to be used as a custom bootstrap method, or as a custom replica creation method.

Now there is need of one more Barman related script in Patroni to handle switching of config models in Barman upon `on_role_change` events.

However, instead of creating another Patroni script, let's say `patroni_barman_config_switch`, and duplicating a lot of logic in the code, we decided to refactor the code so:

* Instead of two separate scripts (`patroni_barman_recover` and `patroni_barman_config_switch`), we have a single script (`patroni_barman`) with 2 sub-commands (`recover` and `config-switch`)

This is the overview of changes that have been performed:

* File `patroni.scripts.barman_recover` has been removed, and its logic has been split into a few files:
  * `patroni.scripts.barman.cli`: handles the entrypoint of the new `patroni_barman` command, exposing the argument parser and calling the appropriate functions depending on the sub-command
  * `patroni.scripts.barman.utils`: implements utilitary enums, functions and classes wich can be used by `cli` and by sub-commands implementation:
    * retry mechanism
    * logging set up
    * communication with pg-backup-api
  * `patroni.scripts.barman.recover`: implements the `recover` sub-command only
* File `patroni.tests.test_barman_recover` has been renamed as `patroni.tests.test_barman`
* File `patroni.scripts.barman.config_switch` was created to implement the `config-switch` sub-command only
* `setup.py` has been changed so it generates a `patroni_barman` application instead of `patroni_barman_recover`
* Docs and unit tests were updated accordingly

References: PAT-154.

### CODE REVIEW

File: patroni/scripts/barman/config_switch.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        except RetriesExceeded:
+            logging.error("Maximum number of retries exceeded, exiting.")
+            sys.exit(ExitCode.HTTP_ERROR)
+
+        if status != OperationStatus.IN_PROGRESS:
+            break
+
+        logging.info("Config switch operation %s is still in progress",
+                     operation_id)
+        time.sleep(5)
```
Comment (CyberDem0n) [inline patroni/scripts/barman/config_switch.py:122]: Hmm, these lines seems to be not reached by unit tests

Comment (barthisrael) [reply patroni/scripts/barman/config_switch.py:122]: Good catch, yes, at a first glance they doesn't seem covered.
Let me take a look at that.

Comment (barthisrael) [reply patroni/scripts/barman/config_switch.py:122]: Just submitted a commit:)
Waiting for the checks to run.

File: patroni/scripts/barman/config_switch.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    if _should_skip_switch(args):
+        logging.info("Config switch operation was skipped (role=%s, "
+                     "switch_when=%s).", args.role, args.switch_when)
+        sys.exit(ExitCode.CONFIG_SWITCH_SKIPPED)
+
+    barman_model, reset = args.barman_model, args.reset
+
+    if all([barman_model, reset]) or not any([barman_model, reset]):
+        logging.error("One, and only one among 'barman_model' ('%s') and "
+                      "'reset' ('%s') should be given", barman_model, reset)
```
Comment (CyberDem0n) [inline patroni/scripts/barman/config_switch.py:144]: I think we could use XOR here :)
```suggestion
    if not bool(args.barman_model) ^ bool(args.reset):
        logging.error("One, and only one among 'barman_model' ('%s') and "
                      "'reset' ('%s') should be given", args.barman_model, args.reset)
```

**Review by CyberDem0n — COMMENTED** [review]
I understand that we already had plenty of sys.exit() calls in the original code, but now I start thinking that it looks a bit ugly, in terms of the actual code and also tests where we have to use `with self.assertRaises(SystemExit)`.

What if we just return the exit code instead of calling sys.exit() and have it just in one place: https://github.com/zalando/patroni/pull/3016/files#diff-014d506e8b5eac5787fa6961cb6c71c1bcf34766832c408df655384848c05f13R225?

IMO it'll make a code and tests a bit cleaner. What do you think @barthisrael?

File: patroni/scripts/barman/config_switch.py (hunk 2 above)
Comment (barthisrael) [reply patroni/scripts/barman/config_switch.py:144]: I like this -- didn't thought of that, much cleaner :)

**Comment by barthisrael** [comment]
> I understand that we already had plenty of sys.exit() calls in the original code, but now I start thinking that it looks a bit ugly, in terms of the actual code and also tests where we have to use `with self.assertRaises(SystemExit)`.
> 
> What if we just return the exit code instead of calling sys.exit() and have it just in one place: https://github.com/zalando/patroni/pull/3016/files#diff-014d506e8b5eac5787fa6961cb6c71c1bcf34766832c408df655384848c05f13R225?
> 
> IMO it'll make a code and tests a bit cleaner. What do you think @barthisrael?

Totally agree, actually.
The more I added those, the more I got uncomfortable -- but then I forgot about that and focused on other parts.
In theory the CLI should handle the "exit", instead of everyone attempting to freaking exit everywhere.

**Comment by barthisrael** [comment]
I need to step back now. I'll rewrite that stuff tomorrow morning and submit a new commit.

File: patroni/scripts/barman/config_switch.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    :param api: a :class:`PgBackupApi` instance to handle communication with
+        the API.
+    :param barman_server: name of the Barman server which config is to be
+        switched.
+    :param barman_model: name of the Barman model to be applied to the server,
+        if any.
+    :param reset: ``True`` if you would like to unapply the currently active
+        model for the server, if any.
+
+    :returns: ``True`` if config was successfully switched, ``False``
```
Comment (hughcapet) [inline patroni/scripts/barman/config_switch.py:88]: not anymore:)

File: patroni/scripts/barman/config_switch.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+
+def run_barman_config_switch(api: "PgBackupApi", args: Namespace) -> int:
+    """Run a remote ``barman config-switch`` through the ``pg-backup-api``.
+
+    :param api: a :class:`PgBackupApi` instance to handle communication with
+        the API.
+    :param args: arguments received from the command-line of
+        ``patroni_barman config-switch`` command.
+    """
```
Comment (hughcapet) [inline patroni/scripts/barman/config_switch.py:136]: :returns: ?
(also in run_barman_recover())

File: patroni/scripts/barman/config_switch.py (hunk 3 above)
Comment (barthisrael) [reply patroni/scripts/barman/config_switch.py:88]: Turns out the guy who is a PITA about docstrings failed with them 😆 
Good catch, I'm fixing both!

File: patroni/scripts/barman/config_switch.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+    if args.switch_when == "promoted" and args.role not in ["master",
+                                                            "primary",
+                                                            "promoted"]:
+        ret = True
+    elif args.switch_when == "demoted" and args.role not in ["replica",
+                                                             "demoted"]:
+        ret = True
+
+    return ret
```
Comment (CyberDem0n) [inline patroni/scripts/barman/config_switch.py:68]:
```suggestion
    if args.switch_when == "promoted":
        return = args.role not in {"master", "primary", "promoted"}
    if args.switch_when == "demoted":
        return args.role not in {"replica", "demoted"}
    return False
```

File: patroni/scripts/barman/config_switch.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        return ExitCode.INVALID_ARGS
+
+    rc: int = _switch_config(api, args.barman_server, args.barman_model, args.reset)
+
+    if rc == ExitCode.CONFIG_SWITCH_DONE:
+        logging.info("Config switch operation finished successfully.")
+    else:
+        logging.error("Config switch operation failed.")
+
+    return rc
```
Comment (CyberDem0n) [inline patroni/scripts/barman/config_switch.py:157]: I guess it is better to have this logging in the `_switch_config()` function.
```suggestion
    return _switch_config(api, args.barman_server, args.barman_model, args.reset)
```

File: patroni/scripts/barman/config_switch.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+            break
+
+        logging.info("Config switch operation %s is still in progress",
+                     operation_id)
+        time.sleep(5)
+
+    if status == OperationStatus.DONE:
+        return ExitCode.CONFIG_SWITCH_DONE
+    else:
+        return ExitCode.CONFIG_SWITCH_FAILED
```
Comment (CyberDem0n) [inline patroni/scripts/barman/config_switch.py:126]:
```suggestion
    if status == OperationStatus.DONE:
        logging.info("Config switch operation finished successfully.")
        return ExitCode.CONFIG_SWITCH_DONE
    else:
        logging.error("Config switch operation failed.")
        return ExitCode.CONFIG_SWITCH_FAILED
```

File: patroni/scripts/barman/recover.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+            break
+
+        logging.info("Recovery operation %s is still in progress",
+                     operation_id)
+        time.sleep(loop_wait)
+
+    if status == OperationStatus.DONE:
+        return ExitCode.RECOVERY_DONE
+    else:
+        return ExitCode.RECOVERY_FAILED
```
Comment (CyberDem0n) [inline patroni/scripts/barman/recover.py:104]:
```suggestion
    if status == OperationStatus.DONE:
        logging.info("Recovery operation finished successfully.")
        return ExitCode.RECOVERY_DONE
    else:
        logging.error("Recovery operation failed.")
        return ExitCode.RECOVERY_FAILED
```

File: patroni/scripts/barman/recover.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    rc = _restore_backup(api, args.barman_server, args.backup_id,
+                         args.ssh_command, args.data_directory,
+                         args.loop_wait)
+
+    if rc == ExitCode.RECOVERY_DONE:
+        logging.info("Recovery operation finished successfully.")
+    else:
+        logging.error("Recovery operation failed.")
+
+    return rc
```
Comment (CyberDem0n) [inline patroni/scripts/barman/recover.py:127]:
```suggestion
    return _restore_backup(api, args.barman_server, args.backup_id,
                           args.ssh_command, args.data_directory,
                           args.loop_wait)
```

**Review by CyberDem0n — COMMENTED** [review]
A couple of nit to have slightly smaller code :)

File: tests/test_barman.py [hunk 10]
```diff
@@ -0,0 +1,766 @@
...
+        mock_log_error.assert_called_once_with("An issue was faced while trying to create a config switch operation: "
+                                               "%r",
+                                               mock_create_op.side_effect)
+        mock_sleep.assert_not_called()
+
+        # get status retries exceeded
+        mock_create_op.reset_mock()
+        mock_create_op.side_effect = None
+        mock_log_error.reset_mock()
+        mock_log_info.reset_mock()
```
Comment (hughcapet) [inline tests/test_barman.py:579]:
```suggestion
```

File: tests/test_barman.py [hunk 11]
```diff
@@ -0,0 +1,766 @@
...
+            BarmanRecoverExitCode.RECOVERY_DONE,
+        )
+
+        mock_create_op.assert_called_once_with(BARMAN_SERVER, BACKUP_ID, SSH_COMMAND, DATA_DIRECTORY)
+        mock_get_status.assert_called_once_with(BARMAN_SERVER, "some_id")
+        mock_log_info.assert_has_calls([
+            mock.call("Created the recovery operation with ID %s", "some_id"),
+            mock.call("Recovery operation finished successfully."),
+        ])
+        mock_log_error.assert_not_called()
```
[the remark this answers is not in the document]
Comment (barthisrael) [reply tests/test_barman.py:300]: Would you like to remove this is "slow restore", "fast config switch", and "slow config switch" as well?

[the remark this answers is not in the document]
Comment (hughcapet) [reply tests/test_barman.py:300]: no, here it is just about the duplicated line:)

[the remark this answers is not in the document]
Comment (barthisrael) [reply tests/test_barman.py:300]: Oh 🤦 I didn't notice this in the GitHub interface, now I see that in the IDE.

**Comment by CyberDem0n** [comment]
:+1:

**Pull request merged**